File: ts/nestjs/src/validation/reservations.ts
import type {
  Reservation,
  ReservationReplay,
} from '../contracts/reservations.contract.js';
import {
  IdempotencyConflict,
  ProductNotFound,
} from '../exceptions/reservations.error.js';

export function validateReplayInput(
  replay: ReservationReplay,
  productId: string,
): Reservation {
  if (replay.storedProductId !== productId) {
    throw new IdempotencyConflict();
  }
  return replay.reservation;
}

export function validateProductExists(productExists: boolean): void {
  if (!productExists) throw new ProductNotFound();
}

File: python/fastapi/src/template_api/validation/reservations.py
from template_api.contracts.reservations import Product, ReplayRecord, Reservation
from template_api.exceptions.reservations import (
    IdempotencyConflict,
    ProductNotFound,
)


def validate_replay_product(
    requested_product_id: str, replay: ReplayRecord
) -> Reservation:
    if replay.product_id != requested_product_id:
        raise IdempotencyConflict()
    return replay.reservation


def validate_product_exists(product: Product | None) -> Product:
    if product is None:
        raise ProductNotFound()
    return product
